File: js/time-math.js
/**
 * Hako — time-math.js
 * Pure date helpers. Mirrors the Python self_buckets/time_math.py.
 *
 * Date convention: ISO date strings (YYYY-MM-DD) everywhere on the boundary.
 * Internal calculations construct Date objects in local time at midnight to
 * avoid timezone surprises.
 */

function todayISO() {
  return toISODate(new Date());
}

function toISODate(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

function parseISODate(iso) {
  // Construct as local-midnight to avoid TZ shifts that occur with `new Date('YYYY-MM-DD')`.
  const [y, m, d] = iso.split('-').map(Number);
  return new Date(y, m - 1, d);
}

function addDays(iso, n) {
  const d = parseISODate(iso);
  d.setDate(d.getDate() + n);
  return toISODate(d);
}

function dayOfWeekISO(iso) {
  // Monday = 0 ... Sunday = 6
  const d = parseISODate(iso);
  return (d.getDay() + 6) % 7;
}

function weekStart(iso) {
  const offset = dayOfWeekISO(iso);
  return addDays(iso, -offset);
}

function weekEnd(iso) {
  return addDays(weekStart(iso), 6);
}

function weekRange(iso) {
  const start = weekStart(iso);
  return [start, addDays(start, 6)];
}

function recentWeekStarts(todayIso, n) {
  if (n < 1) return [];
  const thisMonday = weekStart(todayIso);
  const result = [];
  for (let i = n - 1; i >= 0; i--) {
    result.push(addDays(thisMonday, -7 * i));
  }
  return result;
}

function consecutiveFloorStreak(weeklyCounts, target, mostRecentWeekIso) {
  let streak = 0;
  let cursor = mostRecentWeekIso;
  // Safety bound: don't walk back forever (max 520 = 10 years)
  for (let i = 0; i < 520; i++) {
    const count = weeklyCounts[cursor] || 0;
    if (count >= target) {
      streak += 1;
      cursor = addDays(cursor, -7);
    } else {
      break;
    }
  }
  return streak;
}

function isDrifting(weeklyHours, targetHours, mostRecentWeekIso, opts = {}) {
  const thresholdPct = opts.thresholdPct ?? 0.70;
  const weeksRequired = opts.weeksRequired ?? 2;
  if (targetHours <= 0 || weeksRequired < 1) return false;
  const threshold = targetHours * thresholdPct;
  let cursor = mostRecentWeekIso;
  for (let i = 0; i < weeksRequired; i++) {
    const actual = weeklyHours[cursor] || 0;
    if (actual >= threshold) return false;
    cursor = addDays(cursor, -7);
  }
  return true;
}

function humanDate(iso) {
  const d = parseISODate(iso);
  return d.toLocaleDateString(undefined, { weekday: 'long', day: 'numeric', month: 'short', year: 'numeric' });
}

function shortDate(iso) {
  const d = parseISODate(iso);
  return d.toLocaleDateString(undefined, { day: 'numeric', month: 'short' });
}

function dayShort(iso) {
  const d = parseISODate(iso);
  return d.toLocaleDateString(undefined, { weekday: 'short' });
}

File: self_buckets/time_math.py
"""Pure date/time helpers — no DB dependencies, easy to test."""
from __future__ import annotations

from datetime import date, timedelta


def week_start(d: date) -> date:
    """Return the Monday of the ISO week containing d."""
    return d - timedelta(days=d.weekday())


def week_end(d: date) -> date:
    """Return the Sunday of the ISO week containing d."""
    return week_start(d) + timedelta(days=6)


def week_range(d: date) -> tuple[date, date]:
    """Return (Monday, Sunday) for the ISO week containing d."""
    start = week_start(d)
    return start, start + timedelta(days=6)


def recent_week_starts(today: date, n: int) -> list[date]:
    """Return the last n Mondays (including this week's Monday), oldest first."""
    if n < 1:
        return []
    this_monday = week_start(today)
    return [this_monday - timedelta(weeks=n - 1 - i) for i in range(n)]


def consecutive_floor_streak(
    weekly_counts: dict[date, int],
    target: int,
    most_recent_week: date,
) -> int:
    """
    Count consecutive most-recent weeks that met the floor (count >= target).

    Walks backward from most_recent_week until a week fails the target.
    """
    streak = 0
    cursor = most_recent_week
    while True:
        count = weekly_counts.get(cursor, 0)
        if count >= target:
            streak += 1
            cursor = cursor - timedelta(weeks=1)
        else:
            break
    return streak


def is_drifting(
    weekly_hours: dict[date, float],
    target_hours: float,
    most_recent_week: date,
    threshold_pct: float = 0.70,
    weeks_required: int = 2,
) -> bool:
    """
    Return True if the most recent `weeks_required` weeks all fell below
    `threshold_pct` of the weekly target.
    """
    if target_hours <= 0 or weeks_required < 1:
        return False
    threshold = target_hours * threshold_pct
    cursor = most_recent_week
    for _ in range(weeks_required):
        actual = weekly_hours.get(cursor, 0.0)
        if actual >= threshold:
            return False
        cursor = cursor - timedelta(weeks=1)
    return True
